python/GafferCyclesTest/IECoreCyclesPreviewTest/RendererTest.py

```diff
@@ -2053,5 +2053,175 @@ def testOSLInSVMShadingSystem( self ) :
+				self.assertIn( name, film )
+				self.assertEqual( film[name], value )
+
+		# Remove all our options and check that we get back to the
+		# original defaults.
+
+		for name in options.keys() :
+			renderer.option( name, None )
+
+		film = renderer.command( "cycles:queryIntegrator", {} )
+		for name in options.keys() :
+			with self.subTest( name = name ) :
+				name = name.replace( "cycles:integrator:", "" )
+				self.assertEqual( film[name], defaults[name] )
+
+	def testUnknownOptions( self ) :
+
+		renderer = GafferScene.Private.IECoreScenePreview.Renderer.create(
+			"Cycles",
+			GafferScene.Private.IECoreScenePreview.Renderer.RenderType.Interactive,
+		)
+
+		with IECore.CapturingMessageHandler() as mh :
+			renderer.option( "cycles:invalid", IECore.IntData( 10 ) )
+
+		self.assertEqual( len( mh.messages ), 1 )
+		self.assertEqual( mh.messages[0].level, IECore.Msg.Level.Warning )
+		self.assertEqual( mh.messages[0].message, 'Unknown option "cycles:invalid".' )
+
+	def testThreads( self ) :
+
+		for threads, expectedThreads in [
+			( 1, 1 ),
+			( 2, 2 ),
+			( -1, max( IECore.hardwareConcurrency() -1, 1 ) ),
+			( -10, max( IECore.hardwareConcurrency() -10, 1 ) ),
+			( -10000, 1 ),
+		] :
+
+			with self.subTest( threads = threads ) :
+
+				renderer = GafferScene.Private.IECoreScenePreview.Renderer.create(
+					"Cycles",
+					GafferScene.Private.IECoreScenePreview.Renderer.RenderType.Interactive,
+				)
+
+				renderer.output(
+					"testOutput",
+					IECoreScene.Output(
+						"test",
+						"ieDisplay",
+						"rgba",
+						{
+							"driverType" : "ImageDisplayDriver",
+							"handle" : "testThreads",
+						}
+					)
+				)
+
+				renderer.option( "cycles:session:threads", IECore.IntData( threads ) )
+				## \todo We currently need to do a render for the threads value to
+				# be flushed into the session params. But in future we should be able
+				# to remove that.
+				renderer.render()
+				self.assertEqual( renderer.command( "cycles:querySession", {} )["threads"].value, expectedThreads )
+
+	def testDevices( self ) :
+
+		typeIndices = {}
+		for device in GafferCycles.devices :
+
+			deviceType = device["type"]
+			typeIndex = typeIndices.setdefault( deviceType, 0 )
+			typeIndices[device["type"]] += 1
+
+			with self.subTest( device = device["id"] ) :
+
+				renderer = GafferScene.Private.IECoreScenePreview.Renderer.create(
+					"Cycles",
+					GafferScene.Private.IECoreScenePreview.Renderer.RenderType.Interactive,
+				)
+
+				renderer.output(
+					"testOutput",
+					IECoreScene.Output(
+						"test",
+						"ieDisplay",
+						"rgba",
+						{
+							"driverType" : "ImageDisplayDriver",
+							"handle" : "testThreads",
+						}
+					)
+				)
+
+				renderer.option( "cycles:shadingsystem", IECore.StringData( "SVM" ) )
+				renderer.option( "cycles:device", IECore.StringData( f"{deviceType}:{typeIndex:02d}" ) )
+				## \todo We currently need to do a render for the device to
```

Comment:
Ok this is giving me a hint as to where you're going with this, using the `command` part of the API you could set those session/scene params _before_ any scene building/rendering potentially?

Reply:
I did consider modifying the API so that all the initial options were passed to `Renderer::create()`, but that was looking fiddly on the client side (SceneGadget in particular). So my current plan is to buffer up the options and create the Session the first time we need it - typically when something like `Renderer::object()` is called. Not ideal, but since we always send options before objects I hope it'll work.

I've definitely seen crashes caused by resets done before rendering has started, so I think making one session and sticking with it has to be the plan one way or another.